### PR TITLE
feat: Precharge fn loading gas cost

### PR DESCRIPTION
**Pull request by jakmeier**
The code prepartion steps for executing a smart contract function call
used to be charged after the work has been done. This protocol change
reorders this.

The effect of the procotocol change is that when a function call fails
during contract loading, it will now have a non-zero gas cost.

Incidentally, this change also brings wasmtime failure behavior
during loading in-line with wasmer0 and wasmer2.

Test plan
---------
Tests for this feature must ensure several things.
1. The change does not affect the behavior with disabled feature.
2. Enabled feature has the desired effect of charging gas for
loading code even when contract loading fails on new versions.
3. Future changes do not accidentally make old versions incompatible.

Point 1 is mostly covered by existing tests that already check that
gas is zero during failures of all kinds. These tests on the stable
version are unchanged but the nightly_protocol_feature version checks
that the expected amount of gas is charged.

The test `test_gas_exceed_loading`has been added to also check the
scenario where gas runs out during contract loading.

Upon feature stabilization, these tests will only check the new
behaviour, therefore they cannot cover point 3.

Point 2 and 3 are covered with new tests in the module
`fix_contract_loading_cost_protocol_upgrade`
which compare the results of different control flows in old/new version.

### CODE REVIEW

File: runtime/near-vm-runner/src/tests.rs [hunk 1]
```diff
@@ -136,10 +133,26 @@ fn gas_and_error_match(
             assert_eq!(outcome.burnt_gas, gas, "burnt gas differs");
         }
         None => {
-            assert_eq!(outcome_and_error.0.used_gas, 0);
-            assert_eq!(outcome_and_error.0.burnt_gas, 0);
+            assert_eq!(outcome_and_error.0.used_gas, 0, "used gas non-zero");
+            assert_eq!(outcome_and_error.0.burnt_gas, 0, "burnt gas non-zero");
```
Comment (matklad) [inline runtime/near-vm-runner/src/tests.rs:137]: Wait, why do we even make the `expected_gas` optional here? I think this can be just an `u64` now, after the first error refactor? Might be worth sending as a separate PR before this one, to separate simple change to a large number of tests. 

Also, I really don't like the shape of runtime tests, they are too chunrn-prone. Some day we need to make the tests here more data-driven...

File: runtime/near-vm-runner/src/tests/runtime_errors.rs [hunk 2]
```diff
@@ -873,3 +883,297 @@ fn test_nan_sign() {
...
+                        (module
+                          (type (;0;) (func))
+                          (func (;0;) (type 0)
+                            i32.const 1
+                            i32.const 1
+                            i32.div_s
+                            return
+                          )
+                          (export "main" (func 0))
+                        )"#,
```
Comment (matklad) [inline runtime/near-vm-runner/src/tests/runtime_errors.rs:926]: For new code, let's try to stick with human-readable wat, without cryptic `(;0;)`:

```
(module
  (func (export "main")
    i32.const 1
    i32.const 1
    i32.div_s
    return
  )
)
```

File: runtime/near-vm-runner/src/wasmer2_runner.rs [hunk 3]
```diff
@@ -524,19 +528,6 @@ impl crate::runner::VM for Wasmer2VM {
...
-                MethodResolveError::MethodEmptyName,
-            ));
-            return VMResult::nop_outcome(error);
-        }
-        let artifact =
-            cache::wasmer2_cache::compile_module_cached_wasmer2(code, &self.config, cache);
-        let artifact = match into_vm_result(artifact) {
-            Ok(it) => it,
-            Err(err) => return VMResult::nop_outcome(err),
-        };
```
Comment (matklad) [inline runtime/near-vm-runner/src/wasmer2_runner.rs:538]: Hm, so this case is actually interesting I think. It's not clear to me if we should charge contract loading cost if compilation error happens, as we don't actually load the contract in this case. 

Hypothetically, if, at some point, we add shape-dependent loading costs (based, eg, on the number of fns in the contract), than we'll have to charge this cost *after* compilation. This seems to be cleaner in some way. 

OTOH, the current strategy of just always charging this cost seems *much* simpler. And, even with shape-dependent costs *and* JIT compilation we might want to charge flat cost before we start compiling, and than shaped cost after we parsed WASM.

So, I guess the current approach is a good one, but, still, an interesting find I didn't realise before!

**Review by matklad — COMMENTED** [review]
Did a first-pass review! At a glance, this looks about right! I want to take a closer look at two things:

* `run_with_vm_kind` feels like it probably should be a method of the VM instead -- we sepcifically removed such a function a while ago
* we had to do a bunch of changes to test code -- I wish we didn't have to. I *think* that's the pre-existing effect of our testing infra being fragile, but maybe there's something smart we can do here.


Not approving yet, just wanted to give a first batch of encouraging comments!

File: runtime/near-vm-runner/src/tests/runtime_errors.rs (hunk 2 above)
Comment (jakmeier) [reply runtime/near-vm-runner/src/tests/runtime_errors.rs]: Done

File: runtime/near-vm-runner/src/wasmer2_runner.rs (hunk 3 above)
Comment (jakmeier) [reply runtime/near-vm-runner/src/wasmer2_runner.rs:538]: Yeah, we don't necessarily have to charge the loading fee in case of a compilation error. But I to keep it simple until having an actual good reason to make it more complex.

> even with shape-dependent costs and JIT compilation we might want to charge flat cost before we start compiling

Completely agreed. This is the main justification for me why doing it this way really doesn't seem to hurt.

File: runtime/near-vm-logic/src/logic.rs [hunk 4]
```diff
@@ -2663,6 +2653,7 @@ impl<'a> VMLogic<'a> {
         }
     }
 
+    #[cfg(not(feature = "protocol_feature_fix_contract_loading_cost"))]
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs:2656]: Hm, I mean the gas_counter is a private field of `VMLogic` and probably should remain private. So I don't think I can easily move that code out of this module.

**Comment by jakmeier** [comment]
> Did a first-pass review! At a glance, this looks about right! I want to take a closer look at two things:
> 
>     * `run_with_vm_kind` feels like it probably should be a method of the VM instead -- we sepcifically removed such a function a while ago
> 
>     * we had to do a bunch of changes to test code -- I wish we didn't have to. I _think_ that's the pre-existing effect of our testing infra being fragile, but maybe there's something smart we can do here.
> 
> 
> Not approving yet, just wanted to give a first batch of encouraging comments!

Can you take another look please? :)

My position hasn't change much here: I don't like adding the argument to `VM::run()` (and how we have to touch all tests as a result of it) but I don't see a better way forward right now.

**Comment by jakmeier** [comment]
Thanks to a helpful discussion with @matklad, I think I managed to simplify this PR a lot in terms of files and LOC changed. It should be much easier to review.

Also pinging @nagisa, as I told you before that you should probably wait before reviewing this. I think now it is in a decent state to review.

File: runtime/near-vm-runner/src/wasmtime_runner.rs [hunk 5]
```diff
@@ -225,25 +217,31 @@ impl crate::runner::VM for WasmtimeVM {
...
-        if logic.add_contract_loading_fee(code.code().len() as u64).is_err() {
-            let err = VMError::FunctionCallError(FunctionCallError::HostError(
-                near_vm_errors::HostError::GasExceeded,
-            ));
-            return VMResult::abort(logic, err);
+        let result =
+            logic.before_code_loading(method_name, current_protocol_version, code.code().len());
+        if result.is_err() {
+            return VMResult::abort(logic, result.unwrap_err());
+        }
```
Comment (nagisa) [inline runtime/near-vm-runner/src/wasmtime_runner.rs:224]:
```suggestion
        if let Err(e) = result {
            return VMResult::abort(logic, e);
        }
```

File: runtime/near-vm-runner/src/wasmtime_runner.rs [hunk 6]
```diff
@@ -285,8 +304,15 @@ impl crate::runner::VM for WasmtimeVM {
...
-                    // TODO: This should return an outcome to account for loading cost
-                    VMResult::nop_outcome(err)
+                    if checked_feature!(
+                        "protocol_feature_fix_contract_loading_cost",
+                        FixContractLoadingCost,
+                        current_protocol_version
+                    ) {
+                        return VMResult::abort(logic, err);
+                    } else {
+                        return VMResult::nop_outcome(err);
```
Comment (nagisa) [inline runtime/near-vm-runner/src/wasmtime_runner.rs:314]: Are there any instances where `nop_outcome` can be called with this feature enabled? If not, maybe it would make sense to add an `assert` in it?

File: runtime/near-vm-runner/src/wasmer_runner.rs [hunk 7]
```diff
@@ -342,20 +331,44 @@ impl crate::runner::VM for Wasmer0VM {
...
-        if logic.add_contract_loading_fee(code.code().len() as u64).is_err() {
-            let err = VMError::FunctionCallError(FunctionCallError::HostError(
-                near_vm_errors::HostError::GasExceeded,
-            ));
-            return VMResult::abort(logic, err);
+        let result =
+            logic.before_code_loading(method_name, current_protocol_version, code.code().len());
+        if result.is_err() {
+            return VMResult::abort(logic, result.unwrap_err());
+        }
```
Comment (nagisa) [inline runtime/near-vm-runner/src/wasmer_runner.rs:338]:
```suggestion
        if let Err(e) = result {
            return VMResult::abort(logic, e);
        }
```

File: runtime/near-vm-runner/src/wasmer_runner.rs [hunk 8]
```diff
@@ -342,20 +331,44 @@ impl crate::runner::VM for Wasmer0VM {
...
+            // version do not have gas costs before reaching this code. (Also
+            // see `test_old_fn_loading_behavior_preserved` for a test that
+            // verifies future changes do not counteract this assumption.)
+            Err(err) => return VMResult::abort(logic, err),
+        };
+
+        let result = logic.after_code_loading(current_protocol_version, code.code().len());
+        if result.is_err() {
+            return VMResult::abort(logic, result.unwrap_err());
         }
```
Comment (nagisa) [inline runtime/near-vm-runner/src/wasmer_runner.rs:357]:
```suggestion
        if let Err(e) = result {
            return VMResult::abort(logic, e);
        }
```

File: runtime/near-vm-runner/src/wasmer_runner.rs [hunk 9]
```diff
@@ -342,20 +331,44 @@ impl crate::runner::VM for Wasmer0VM {
...
-            return VMResult::nop_outcome(e);
+            if checked_feature!(
+                "protocol_feature_fix_contract_loading_cost",
+                FixContractLoadingCost,
+                current_protocol_version
+            ) {
+                return VMResult::abort(logic, e);
+            } else {
+                return VMResult::nop_outcome(e);
+            }
```
Comment (nagisa) [inline runtime/near-vm-runner/src/wasmer_runner.rs:371]: Would it be worthwhile perhaps to abstract this pattern into a method? It gets repetitive fast, and I don't think there's much more value in open coding it vs something like

```rust
VMResult::abort_but_nop_outcome_in_old_protocol(protocol_version, logic, e)
```

File: runtime/near-vm-runner/src/wasmer2_runner.rs (hunk 3 above)
Comment (nagisa) [reply runtime/near-vm-runner/src/wasmer2_runner.rs:530]: Given that we can't be sure if a `FunctionCall` will result in a compilation or not, and we AFAIU don't want conditionally or unconditionally charge the possibly very significant fees for every function invocation, it makes complete sense to me to just charge the “executable loading” fee before compilation.

File: runtime/near-vm-runner/src/tests/runtime_errors.rs [hunk 10]
```diff
@@ -873,3 +883,295 @@ fn test_nan_sign() {
...
+            gas_and_error_match(
+                res,
+                expected_gas,
+                Some(VMError::FunctionCallError(FunctionCallError::CompilationError(
+                    expected_error.clone(),
+                ))),
+            );
+        }
+
+        {
```
Comment (nagisa) [inline runtime/near-vm-runner/src/tests/runtime_errors.rs:1069]: I'd just split this into two `#[test]`s. There's little shared setup here and having more granular tests makes it easier to see which one fails when they do fail.

File: runtime/near-vm-runner/src/wasmtime_runner.rs (hunk 6 above)
Comment (jakmeier) [reply runtime/near-vm-runner/src/wasmtime_runner.rs:314]: Yes, there are several places we still call `nop_outcome`. Perhaps the most important location is in `runtime/runtime/src/actions.rs` in the body of `execute_function_call()`.

https://github.com/near/nearcore/blob/0775622fc2cd6cd428960b4cd10867e50fef90a9/runtime/runtime/src/actions.rs#L39-L67

Hopefully, I will eventually remove all of these.

File: runtime/near-vm-logic/src/logic.rs [hunk 11]
```diff
@@ -2689,6 +2697,61 @@ impl<'a> VMLogic<'a> {
         let new_used_gas = self.gas_counter.used_gas();
         self.gas_counter.process_gas_limit(new_burn_gas, new_used_gas)
     }
+
+    /// Perform VM independent checks that happen after the instantiation of
+    /// VMLogic but before loading the code. This includes pre-charging gas
+    /// costs for loading the code.
+    pub fn before_code_loading(
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs]: Good point, the naming is indeed confusing. I like "executable" to make it clearer.

File: runtime/near-vm-logic/src/logic.rs [hunk 12]
```diff
@@ -2689,6 +2697,62 @@ impl<'a> VMLogic<'a> {
...
+        &mut self,
+        current_protocol_version: u32,
+        wasm_code_bytes: usize,
+    ) -> std::result::Result<(), VMError> {
+        if !checked_feature!(
+            "protocol_feature_fix_contract_loading_cost",
+            FixContractLoadingCost,
+            current_protocol_version
+        ) {
+            if self.add_contract_loading_fee(wasm_code_bytes as u64).is_err() {
```
Comment (akhi3030) [inline runtime/near-vm-logic/src/logic.rs:2746]: We are not actually looking at what the error returned was and always assuming that it is `GasExceeded`.  I wonder if we can do something better here.  Although this code is probably not part of this PR and probably already existed so does not need to fixed in this PR.

File: runtime/near-vm-runner/src/tests.rs [hunk 13]
```diff
@@ -55,34 +55,45 @@ fn create_context(input: Vec<u8>) -> VMContext {
...
-            &mut fake_external,
-            context,
-            &fees,
-            &promise_results,
-            LATEST_PROTOCOL_VERSION,
-            None,
-        )
-        .outcome_error()
+
+    if let Some(runtime) = vm_kind.runtime(config) {
```
Comment (akhi3030) [inline runtime/near-vm-runner/src/tests.rs:81]: nit: Should we consider keeping the previous style namely because it has fewer indentations?

File: runtime/near-vm-runner/src/tests.rs [hunk 14]
```diff
@@ -134,3 +133,14 @@ fn gas_and_error_match(
     assert_eq!(outcome.burnt_gas, expected_gas, "burnt gas differs");
     assert_eq!(outcome_and_error.1, expected_error);
 }
+
+/// Small helper to compute expected loading gas cost charged before loading.
+/// Includes some hard-coded parameter values that would have to be updated
+/// if they change in the future.
```
Comment (akhi3030) [inline runtime/near-vm-runner/src/tests.rs:139]: Can we add a brief comment as to how these hardcoded values were derived please?  To help newbies like myself 😄

**Review by akhi3030 — COMMENTED** [review]
Please feel free to ignore all my suggestions if you don't think they are handy.  I am just trying my hand at reviewing.

File: runtime/near-vm-runner/src/tests/runtime_errors.rs (hunk 10 above)
Comment (jakmeier) [reply runtime/near-vm-runner/src/tests/runtime_errors.rs:1069]: I am not quite understanding the suggestion.

I mean, yes, I can easily split it into two checker functions. But how to split the actual test `test_fn_loading_gas_protocol_upgrade_fail_preparing`? In one way or another, I will have to duplicate the list of (contract,expected_error) pairs. I'm unable to make this work without having horrible duplication at one place or another... But maybe I just misunderstood the idea.

Comment (nagisa) [reply runtime/near-vm-runner/src/tests/runtime_errors.rs:1069]: Aha, never mind, I missed that this is a helper function not an actual test.

File: runtime/near-vm-logic/src/logic.rs [hunk 15]
```diff
@@ -2673,6 +2675,12 @@ impl<'a> VMLogic<'a> {
         }
     }
 
+    /// Add a cost for loading the contract code in the VM.
+    ///
+    /// This cost is "dumb", in that it does not consider the structure of the
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs]: Yeah I rephrased it to something that (hopefully) sounds a bit more elegant :)

File: runtime/near-vm-logic/src/logic.rs [hunk 16]
```diff
@@ -2673,6 +2675,12 @@ impl<'a> VMLogic<'a> {
...
+    /// contract code, only the size. This is currently the only loading fee.
+    /// A smarter fee could be added, although that would have to happen slightly
+    /// later, as this dumb fee can be pre-charged without looking at the code.
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs]: Later, as in after the loading. Because we need the information from the loading phase to know the structure.

I rephrased it, I hope it is clearer now.

File: runtime/near-vm-logic/src/logic.rs [hunk 17]
```diff
@@ -2689,6 +2697,62 @@ impl<'a> VMLogic<'a> {
         let new_used_gas = self.gas_counter.used_gas();
         self.gas_counter.process_gas_limit(new_burn_gas, new_used_gas)
     }
+
+    /// VM independent setup before performing executable loading.
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs]: done

File: runtime/near-vm-logic/src/logic.rs (hunk 12 above)
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs:2747]: Yeah I see what you mean.

I think passing on the error from inside would look cleaner on the surface. But we would have to consider carefully if this is an observable change. From a quick glance, it is not obvious to me that it could not also be a `HostError::IntegerOverflow`.

And the second consideration would be, if we actually want to return more specific host errors, or if we specifically want to reduce it to just "gas exceeded"...

So yeah, not something I want to tie together with this PR.

File: runtime/near-vm-runner/src/tests.rs (hunk 14 above)
Comment (jakmeier) [reply runtime/near-vm-runner/src/tests.rs]: Yes, I elaborated it some more. But the key here is that `parameter` is a technical term. It is a runtime parameter that is part of protocol configuration.

**Review by matklad — APPROVED** [review]
LGTM!

File: runtime/near-vm-logic/src/logic.rs (hunk 12 above)
Comment (akhi3030) [reply runtime/near-vm-logic/src/logic.rs:2747]: I raised a GH issue about it.  I'll resolve the conversation.
